teps/0086-changing-the-way-result-parameters-are-stored.md

```diff
@@ -0,0 +1,341 @@
+    - Concern around having to create a new ConfigMap**, Role and RoleBinding per TaskRun, when at the end of the day we don't actually care about updating that ConfigMap**, but the TaskRun's results.
+  - This option won't 'scale fail' which would happen with self definition update when the definition itself starts to get too big in unrelated areas (aggregate limit).
+
+2. CRD
+
+  - Would limit the chance of editing with `kubectl edit cm <results>`
+  - Similar benefits to Configmap from a Role and Rolebinding perspective
+  - Webhook to validate the write once immutability
+
+3. A dedicated HTTP Service
+
+  - Potential Auth and HA problems
+  - Could run as part of the controller
+  - Could be a separate HTTP server(s) which write to TaskRuns (or even config maps); task pods connect to this server to submit results, this server records the results (means result size would be limited to what can be stored in the CRD but there probably needs to be an upper bound on result size anyway)
+
+4. Self-update / mutate the TaskRun via admission controller (with the various controls i.e. first write, subsequent read only)
+
+  - Potential issue with self updating its own Status
+
+5. Separate Database
+
+  - Introducing an additional database requirement to Tekton to support the storage of information outside of etcd.
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+## Design Evaluation
+<!--
+How does this proposal affect the reusability, simplicity, flexibility 
+and conformance of Tekton, as described in [design principles](https://github.com/tektoncd/community/blob/master/design-principles.md)
+-->
+
+## Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out?  These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+There is the current alternative of storing result parameters as data in a workspace, however Workspaces
+- require there has to be a storage mechanism in the cluster that can be shared between Tasks. That can be complex, or have performance issues in itself if using an As A Service that orders the storage at spin-up time. Or forces Tasks to all run on the same node. etc. Storage is a big complex adoption hurdle.
+- changes the way end users refer to the result parameter or pass between containers
```

Comment:
I 100% we should not change the interface for retrieving results. 
For producing results, if we decide to provide a new interface, we should maintain backward compatibility anyways.

Since tasks can be written in any programming language, it feels to me that the current file-system based interface works well across languages, and it can be used without the need of extra dependencies.

I wonder if it would make sense to provide a pluggable interface on Tekton side?
We could keep both read and write interfaces from Tasks as they are, provide a default storage implementation as Tekton and let users plug-in other kind of storage based on their requirements (size, speed).